### PR TITLE
Library review and improvements

### PR DESCRIPTION
**Pull request by visciang**
Hello @corka149,
thank you for the library!

We have evaluated the library while working on some JSON API related features.
During the evaluation we have fixed some bugs, extended some functionalities, etc.

The result is a major (2.0) jump, since some API signature have been changed, check out the CHANGELOG
for a summary.

We are currently using our internal fork of the library but we would like to contribute back.
Let us know if this can be accepted or not.

Thank you,
the Athonet Team ❤️ ❤️ ❤️

### CODE REVIEW

**Comment by corka149** [comment]
Hey @visciang . Thanks for the feedback and huge effort!

I am currently short on time and will try to have a look at it on the weekend.

File: lib/jsonpatch/utils.ex [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+      do_update_destination(target, value, [], fragments, opts)
+    end
+  end
+
+  @doc """
+  Unescape `~1` to  `/` and `~0` to `~`.
+  """
+  @spec unescape(fragment :: String.t() | integer()) :: String.t()
+  def unescape(fragment) when is_binary(fragment) do
+    fragment
```
Comment (corka149) [inline lib/jsonpatch/utils.ex:186]: The `unescape` should contain a check if the replacing is necessary. See https://github.com/corka149/jsonpatch/issues/11

File: lib/jsonpatch/utils.ex [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    |> String.replace("~0", "~")
+    |> String.replace("~1", "/")
+  end
+
+  @doc """
+  Escape `/` to `~1 and `~` to `~0`.
+  """
+  @spec escape(fragment :: String.t() | integer()) :: String.t()
+  def escape(fragment) when is_binary(fragment) do
```
Comment (corka149) [inline lib/jsonpatch/utils.ex:195]: Here also, the `escape` should contain a check if the replacing is necessary. See https://github.com/corka149/jsonpatch/issues/11

File: lib/jsonpatch_exception.ex [hunk 3]
```diff
@@ -1,14 +0,0 @@
-defmodule JsonpatchException do
```
Comment (corka149) [inline lib/jsonpatch_exception.ex]: Why was it exchange against the error as struct?

File: CHANGELOG [hunk 4]
```diff
@@ -1,3 +1,19 @@
...
+- Change - MOVE operation to be based on on COPY+REMOVE operation (as per RFC)
+- Change - REPLACE operation to be based on on REMOVE+ADD operation (as per RFC)
+- Change - `Jsonpatch.apply_patch()` signature changes:
+  - patches can be defined as `Jsonpatch.Operation.Add/Copy/Remove/...` structs or with plain map conforming to the jsonpatch schema
+  - error reason is now defined with a `{:error, %Jsonpatch.Error{}}` tuple.
+    %Jsonpatch.Error{patch_index: _, path: _, reason: _} reports the patch index, the path and the reason that caused the error.
+- Removed - `Jsonpatch.Mapper` module, in favour of new Jsonpatch.apply_patch signature
+- Removed - `JsonpatchException` exception
+- Removed - `Jsonpatch.Operation` protocol
+- Feature - introduced new `Jasonpatch.apply_patch()` option `keys: {:custom, convert_fn}` to convert path fragments with a user specific logic
```
Comment (corka149) [inline CHANGELOG]: Typo? (I was actually searching for :jason :laughing: )

File: mix.exs [hunk 5]
```diff
@@ -6,46 +6,41 @@ defmodule Jsonpatch.MixProject do
...
-
-      # DEV
-      ## testing with real json files
-      {:poison, "~> 4.0", only: [:test]},
-      ## code test coverage
-      {:excoveralls, "~> 0.15.3", only: [:test]},
-      ## linting
-      {:credo, "~> 1.6.0", only: [:dev, :test], runtime: false},
-      ## type checking
+      {:jason, "~> 1.0", only: [:dev, :test]},
```
Comment (corka149) [inline mix.exs]: :jason isn't use anywhere or did I miss something?

**Review by corka149 — CHANGES_REQUESTED** [review]
Lots of changes :sweat_smile: . I iterated over it the first time.

File: lib/jsonpatch/utils.ex (hunk 1 above)
Comment (visciang) [reply lib/jsonpatch/utils.ex:186]:
```elixir
:timer.tc(fn ->
  Enum.each(1..1_000_000, fn _ -> String.replace("uhuhuhuhuhuhuhuhuhuhuhuhuhuhhuhuhu", "~0", "/") end)
end)
{1445115, :ok}

:timer.tc(fn ->
  Enum.each(1..1_000_000, fn _ -> String.contains?("uhuhuhuhuhuhuhuhuhuhuhuhuhuhhuhuhu", "~0") end)
end)
{1114837, :ok}
```

So `String.replace` is about 30% slower that  `String.contains?` (in this toy example, on my machine).
But using `String.contains?` imply a second scan with `String.replace`, in the worst case (ie if the string should be escaped). In this case we will pay 1114837 + 1445115 -> 70% slower than single `String.replace`.

So this kind of optimisation depend on the probability distribution of the input string being in the set of string that need an escape or not.

IMHO favouring one and saving 30% or the other and saving 70% is quite arbitrary.

BTW I'm OK to revert, in our use case the impact of the library processing time is very small whatever JSON we patch.

PS: the change was not done with this reasoning, it was just a spontaneous code simplification at the time of review with my colleagues ;).

File: lib/jsonpatch_exception.ex (hunk 3 above)
Comment (visciang) [reply lib/jsonpatch_exception.ex]: Sorry, my bad.
Restored JsonpatchException (instead of a RuntimeException)

**Review by corka149 — APPROVED** [review]
I like the changes.

The only nitpicking I have is about naming a module only `Util`. The name is always an invitation for placing unrelated functions in one module. But I am fine for now with this as the scope of the lib is very small.

**Comment by visciang** [comment]
@corka149 may I ask you to push 2.0.0 on hex.pm?

Thank you
Giovanni

**Comment by corka149** [comment]
@visciang already done ;-)

**Pull request merged**